### PR TITLE
Add matcher callback operations to the transform dialect

### PR DESCRIPTION
**Pull request by ftynse**
These operations allow one to call back to a C++ function to perform a
more advanced form of pattern matching than what is currently available
in the transform dialect or in PDL. They are a temporary scaffolding to
enable end-to-end matching and transformation while we are building out
the matching part of the transform dialect or improve PDL.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp [hunk 1]
```diff
@@ -229,6 +229,48 @@ static Value distributeVectors(ImplicitLocOpBuilder &b, Value funcH,
...
+  if (!hasLeadingEltwise) {
+    result.splitFillH = splitReductionTransformOp.getFillOp();
+    result.splitLinalgH = splitReductionTransformOp.getSplitLinalgOp();
+    result.combinerH = splitReductionTransformOp.getCombiningLinalgOp();
+    return result;
+  }
+
+  auto funcH = b.create<MatchOp>(variantH, func::FuncOp::getOperationName());
+  auto applyPatterns = b.create<ApplyPatternsOp>(funcH, /*rankReducing=*/false);
+  applyPatterns->setAttr(applyPatterns.getBubbleCollapseExpandAttrName(),
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:281]: This is unfortunate, created a bug

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp [hunk 2]
```diff
@@ -229,6 +229,48 @@ static Value distributeVectors(ImplicitLocOpBuilder &b, Value funcH,
...
+  Value combinerH;
+  /// Handle to the original fill operation, may be null if the operation was
+  /// not re-matched.
+  Value originalFillH;
+};
+
+/// Builds transform IR requesting to bubble up the "expand_shape" operation
+/// produced as parent of reduction splitting if necessary for fusion of the
+/// leading elementwise operation.
+static ReductionSplitResult buildExpansionBubbleUp(
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:267]: Can we split this in 2? 
There is a generic builder of expansionBubbleOp that is independent of the ops involved.

Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:267]: There is such a thing as a generic builder of expansionBubbleOp that follows the same API as other things in this file (e.g. Bufferize) and that is independent of the ops involved. 

I was going to ask to split this in 2 between the 2-line op-independent part and the op-specific part OTOH, there is a bigger picture item that ApplyPatternsOp should be able to preserve handles at least in certain cases.

Can you add a TODO? 
Added a bulletpoint to address this later.

**Review by nicolasvasilache — APPROVED** [review]
Last commit LGTM, can you send it upstream directly ?

File: compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp (hunk 2 above)
Comment (ftynse) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:267]: Sure, but not in this PR.

Comment (ftynse) [reply compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp:267]: The preservation is orthogonal and the splitting is introduced in another PR.

File: compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp [hunk 3]
```diff
@@ -993,5 +995,107 @@ DiagnosedSilenceableFailure transform_dialect::ConfigExtractPart::apply(
...
+/// operations associated with operand handles unless they have the
+/// "test.iree_transform_do_not_match" attribute, in which case produces a
+/// silenceable failure.
+static DiagnosedSilenceableFailure testMatchCallbackCallback(
+    transform_dialect::MatchCallbackResult &res, Location loc,
+    const transform::TransformState &state, ValueRange handles) {
+  bool hadFailures = false;
+  for (Value handle : handles) {
+    if (llvm::any_of(state.getPayloadOps(handle), [](Operation *op) {
+          return op->hasAttr("test.iree_transform_do_not_match");
```
[the remark this answers is not in the document]
Comment (ftynse) [reply compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp:1012]: This is specific to testing behavior of a callback in the infra...

**Pull request merged**